### ISSUE TITLE
Resolve Pitest Issues - TreeWalker

### ISSUE THREAD

**Issue by timurt** [checkstyle/checkstyle#10053]
Child Issue of #7797

**Comment by romani** [comment]
All violation should be tried to covered by functional tests, not a UTs. Coverage  by pure UTs is absolute last resort.

**Comment by shashwatj07** [comment]
We are talking about:
https://github.com/checkstyle/checkstyle/blob/578a97541eb50f27db3f5d8f4d9b16d09d8618c6/.ci/pitest.sh#L161-L163

The context is:
https://github.com/checkstyle/checkstyle/blob/6cbc1dc3c01385426b8bcfd9988e92acdc9390a9/src/main/java/com/puppycrawl/tools/checkstyle/TreeWalker.java#L153-L167

1. Mutation is surviving on line 153 since there will be no difference in any functionality even if we walk the AST with an empty ```ordinaryChecks```
2. Same as above for ```commentChecks``` on line 156.
3. On line 160, if filters is empty, there will be absolutely no difference if we ```getFilteredViolations``` (which shall return an empty set) and add them to our violations.

Hence, these checks are redundant and may be removed.

**Comment by rnveach** [comment]
Check on line 153/156 has a purpose. It is a guard to prevent walking an entire AST when it is not needed. It is just going to waste CPU cycles doing something that we know doesn't need to be done. 156 does the extra guarding of rebuilding the AST to bring comments in, which again doesn't need to be done if there is no comment checks.

We would need to gain an understanding how much time these guards help save if they are just to be removed.

**Comment by timurt** [comment]
@shashwatj07 I agree with @rnveach 

Lets try another approach,

For example for this mutation
```
            if (filters.isEmpty()) {
                addViolations(violations);
            }
            else {
                final SortedSet<Violation> filteredViolations =
                    getFilteredViolations(file.getAbsolutePath(), contents, rootAST);
                addViolations(filteredViolations);
            }
```

Maybe we can provide mock check and mock `TreeWalkerFilter` filter to `TreeWalker` and then call `TreeWalker.process(File file, FileText fileText)`

smth like
```
TreeWalker tw = new TreeWalker();
tw.setupChild(mockCheck);
assertThat(tw.process(file, fileText)).size()).is(0);

tw.setupChild(mockFilter);
assertThat(tw.process(file, fileText).size()).is(1);
```

please try it and lets discuss it

**Comment by shashwatj07** [comment]
> ```
> TreeWalker tw = new TreeWalker();
> tw.setupChild(mockCheck);
> assertThat(tw.process(file, fileText)).size()).is(0);
> 
> tw.setupChild(mockFilter);
> assertThat(tw.process(file, fileText).size()).is(1);
> ```
> 
> please try it and lets discuss it

@timurt Will this kill the mutation? I don't think so. The mutation survives on converting ```filters.isEmpty()``` to ```false```. Despite ```filters``` being empty, the if and the else clause both have the same result on the violations.

**Comment by timurt** [comment]
@shashwatj07 

> Will this kill the mutation? I don't think so. The mutation survives on converting filters.isEmpty() to false. Despite filters being empty, the if and the else clause both have the same result on the violations.

we can provide own mock filter, that will throw out violation here `getFilteredViolations(file.getAbsolutePath(), contents, rootAST);`
so when `filters.isEmpty()` is mutated, method `AbstractFileSetCheck. process` will return violation instead of empty list- thats idea

Lets try and see

**Comment by romani** [comment]
No need for mock Checks, just reuse any existing Checks.

**Comment by timurt** [comment]
@romani @shashwatj07 I was wrong with mock objects, tested locally

as @shashwatj07 coded:
```
            if (filters.isEmpty()) {
                addViolations(violations);
            }
            else {
                final SortedSet<Violation> filteredViolations =
                    getFilteredViolations(file.getAbsolutePath(), contents, rootAST);
                addViolations(filteredViolations);
            }
```

can be simplified to, this seems a working approach:
```
                final SortedSet<Violation> filteredViolations =
                    getFilteredViolations(file.getAbsolutePath(), contents, rootAST);
                addViolations(filteredViolations);
```

this will add a little overhead of creating objects and violations iteration

**Comment by pbludov** [comment]
Fixed by #10932
